x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sequence/CircuitBreakerTests.java

```diff
@@ -153,6 +198,284 @@ public void testCircuitBreakerSequnceMatcher() {
         assertEquals("sequence_inflight", e.getMessage());
     }
 
+    public void testMemoryClearedOnSuccessfulRequest() {
+        assertMemoryCleared(2, PIT_ID_TYPE.SUCCESS_ID);
+    }
+
+    public void testMemoryClearedOnShardsException() {
+        assertMemoryCleared(stages, PIT_ID_TYPE.FAILURE_ID);
+    }
+
+    private void assertMemoryCleared(int sequenceFiltersCount, PIT_ID_TYPE failureId) {
+        List<BreakerSettings> eqlBreakerSettings = Collections.singletonList(
+            new BreakerSettings(
+                CIRCUIT_BREAKER_NAME,
+                CIRCUIT_BREAKER_LIMIT,
+                CIRCUIT_BREAKER_OVERHEAD,
+                CircuitBreaker.Type.MEMORY,
+                CircuitBreaker.Durability.TRANSIENT
+            )
+        );
+        try (
+            CircuitBreakerService service = new HierarchyCircuitBreakerService(
+                Settings.EMPTY,
+                eqlBreakerSettings,
+                new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS)
+            );
+            PitMockClient esClient = new PitMockClient(getTestName(), service.getBreaker(CIRCUIT_BREAKER_NAME), PIT_ID_TYPE.FAILURE_ID)
```

Comment:
That specific line you added the comment for should in fact be (as you spotted earlier that `failureId` is not used) `PitMockClient esClient = new PitMockClient(getTestName(), service.getBreaker(CIRCUIT_BREAKER_NAME), failureId)`.

I found that the value of the pit ID is simple and flexible enough to allow me to run two different tests (success and failure) and, also, to control the behavior of the client response in multiple requests for one of the pid ID types.

Can you detail and/or suggest more specific code changes to use a flag or different (client, I assume) instances?

Reply:
What I'm mostly thinking of is to separate the various concerns packed into the PIT ID right now:
- What type of client is used by the test (failing vs. succeeding)
- Keeping track of the client state (how many requests have been answered)
- Being the PIT ID itself

Instantiating `PitMockClient` with a `succeeding` flag (or using two different implementations) can address the first concern. I'm not entirely sure about the client state but that can probably be addressed by a field in the client. This would also resolve the issue mentioned in https://github.com/elastic/elasticsearch/pull/84451/files#r816743228.